source/routing/raptor_api.cpp

```diff
@@ -1082,6 +1103,85 @@ void make_pt_response(navitia::PbCreator& pb_creator,
     }
 }
 
+void filter_direct_path(RAPTOR::Journeys& journeys)
+{
+    journeys.remove_if([](const Journey& j){return !j.is_pt();});
+}
+
+/**
+    Check if a journey is way later than another journey
+
+    Then, we check for each journey the difference between the
+    requested datetime and the arrival datetime (the other way around
+    for non clockwise)
+
+    requested dt
+    *
+                    |=============>
+                          journey2
+
+                                            |=============>
+                                                 journey1
+
+    |-----------------------------|
+       journey2 pseudo duration
+
+    |------------------------------------------------------|
+            journey1 pseudo duration
+ */
+bool way_later(const Journey & j1, const Journey & j2,
+               const NightBusFilter::Params & params)
+{
+    auto & requested_dt = params.requested_datetime;
+    auto & clockwise = params.clockwise;
+
+    auto get_pseudo_duration = [&](const Journey & j) {
+        auto dt = clockwise ? j.arrival_dt : j.departure_dt;
+        return std::abs(dt - requested_dt);
+    };
+
+    auto j1_pseudo_duration = get_pseudo_duration(j1);
+    auto j2_pseudo_duration = get_pseudo_duration(j2);
+
+    auto max_pseudo_duration = j2_pseudo_duration * params.max_factor + params.base_factor;
+
+    return j1_pseudo_duration > max_pseudo_duration;
+}
+
+void filter_late_journeys(RAPTOR::Journeys & journeys,
+                          const NightBusFilter::Params & params)
+{
+    if(journeys.size() == 0)
+        return;
+
+    auto journeys_pairs_gen = utils::make_pairs_generator(journeys);
+    std::vector<RAPTOR::Journeys::const_iterator> late_journeys;
+
+    for(auto journey_pair : journeys_pairs_gen) {
```

Comment:
JUST FOR FUN & TL;DR 

I think what you need here is a Permutation( your generator gives a Combination actually, you did a Permutation later by swapping the pair in the loop, this saves you from detecting self-repeating pair I think :smile: ), the permutation can be done by a cartesian product without repeat, a little search gave me this: https://stackoverflow.com/a/16688622/1614576, this solution handles the problem nicely and generically.
 (I prefer your implementation here which is much simpler to read :smile: )

Reply:
it doesn't handle the self repeating cases...

Reply:
Yes you're right, I've chosen Combination over Permutation to avoid evaluating the swapped case when the first case succeed. 

For fun, it would be nice to generalise `pairs_generator` for a combination of n elements and submit it to boost :p